### PR TITLE
Refactor thread model to be created from the root event

### PR DESCRIPTION
**Pull request by gsouquet**
To review alongside https://github.com/matrix-org/matrix-react-sdk/pull/7670

- Refactor thread model to be created from the root event
- Integrate threads with the /relations API

---
This PR currently has no changelog labels, so will not be included in changelogs.

Add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is plus `X-Breaking-Change` if it's a breaking change.

### CODE REVIEW

File: src/models/thread.ts [hunk 1]
```diff
@@ -91,85 +99,111 @@ export class Thread extends TypedEventEmitter<ThreadEvent> {
...
     /**
      * Add an event to the thread and updates
      * the tail/root references if needed
      * Will fire "Thread.update"
      * @param event The event to add
      */
     public async addEvent(event: MatrixEvent, toStartOfTimeline = false): Promise<void> {
-        if (this.timelineSet.findEventById(event.getId())) {
-            return;
+        // When there's no server side support we man
```
Comment (t3chguy) [inline src/models/thread.ts:113]: we man what?

File: src/models/thread.ts [hunk 2]
```diff
@@ -91,85 +99,111 @@ export class Thread extends TypedEventEmitter<ThreadEvent> {
...
+            await this.client.decryptEventIfNeeded(event, {});
         }
 
-        if (!this.root) {
-            if (event.isThreadRelation) {
-                this.root = event.threadRootId;
-            } else {
-                this.root = event.getId();
+        if (this.hasServerSideSupport && this.initialEventsFetched) {
+            if (event.getTs() > this.lastReply().getTs() && !this.findEventById(event.getId())) {
```
Comment (t3chguy) [inline src/models/thread.ts:137]: relying on timestamps between events for ordering is a no-go given different servers will have different clocks

File: src/models/thread.ts [hunk 3]
```diff
@@ -91,85 +99,111 @@ export class Thread extends TypedEventEmitter<ThreadEvent> {
...
-
         const isThreadReply = event.getRelation()?.rel_type === RelationType.Thread;
         // If no thread support exists we want to count all thread relation
         // added as a reply. We can't rely on the bundled relationships count
         if (!this.hasServerSideSupport && isThreadReply) {
             this.replyCount++;
         }
 
-        if (!this.lastEvent || (isThreadReply && event.getTs() > this.lastEvent.getTs())) {
+        if (!this.lastEvent || (isThreadReply && event.getTs() > this.replyToEvent.getTs())) {
```
Comment (t3chguy) [inline src/models/thread.ts:159]: same as above

File: src/models/thread.ts (hunk 2 above)
Comment (gsouquet) [reply src/models/thread.ts:137]: What would be the alternative? How can I know whether a message is newly incoming?

Comment (gsouquet) [reply src/models/thread.ts]: I'm planning to merge this implementation using the `age`
It has cases in which is could cause the count to drift compared to the value returned by the HS. But this is never going to happen pre-release as we're going to use two homeservers with relatively synced clocks.
I'll create an issue to revisit that before the launch and will see whether we can find a better approach

**Review by t3chguy — APPROVED** [review]

**Pull request merged**